File: codeforces/1900~1999/1901/d.cc
#include <bits/stdc++.h>

using namespace std;

using i64 = long long;

void solve() {
  int n = 0;
  cin >> n;
  vector<i64> a(n+1);
  for (int i = 1; i <= n; ++i) cin >> a[i];

  vector<i64> prefix(n+2), suffix(n+2);
  for (int i = 1; i <= n; ++i) {
    prefix[i] = max(prefix[i], a[i] + (n-i));
    suffix[i] = max(suffix[i], a[i] + i-1);
  }
  for (int i = 1; i <= n; ++i) {
    prefix[i] = max(prefix[i], prefix[i-1]);
  }
  for (int i = n; i >= 1; --i) {
    suffix[i] = max(suffix[i], suffix[i+1]);
  }
  i64 ans = numeric_limits<i64>::max();
  for (int i = 1; i <= n; ++i) {
    // cout << a[i] << ", " << prefix[i-1] << ", " << suffix[i+1] << endl;
    i64 max_val = max({a[i], prefix[i-1], suffix[i+1]});
    ans = min(ans, max_val);
  }
  cout << ans << endl;
}

int main(void) {
  ios::sync_with_stdio(false);
  cin.tie(0);

  solve();
  return 0;
}

File: codeforces/1900~1999/1901/d.rs
use std::io::{stdin, stdout, BufWriter, Write};
use std::str::FromStr;
use std::fmt::Debug;
use std::cmp::{max, min};

fn read_int<T: FromStr>() -> T where <T as FromStr>::Err: Debug {
    let mut s = String::new();
    stdin().read_line(&mut s).unwrap();
    s.trim().parse::<T>().expect("invalid input")
}

fn read_vec<T: FromStr>() -> Vec<T> where <T as FromStr>::Err: Debug {
    let mut s = String::new();
    stdin().read_line(&mut s).unwrap();
    s.split(" ").map(|x| x.trim().parse().expect("invalid input")).collect()
}

fn solve() {
    let stdout = stdout();
    let mut writer = BufWriter::new(stdout.lock());

    let n = read_int::<usize>();
    let mut a = read_vec::<i64>();
    a.insert(0, 0);

    let mut prefix = vec![0; n+2];
    let mut suffix = vec![0; n+2];
    for i in 1..=n {
        prefix[i] = max(prefix[i-1], a[i] + (n-i) as i64);
    }
    for i in (1..=n).rev() {
        suffix[i] = max(suffix[i+1], a[i] + (i-1) as i64);
    }
    let mut ans = i64::MAX;
    for i in 1..=n {
        let start = max(max(suffix[i+1], prefix[i-1]), a[i]);
        ans = min(ans, start);
    }
    writeln!(writer, "{}", ans).ok();
}

fn main() {
    solve();
}
